source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -869,6 +873,61 @@ ServerContextImpl::ServerContextImpl(Stats::Scope& scope,
   }
 }
 
+void ServerContextImpl::populateServerNamesMap(TlsContext& ctx, int pkey_id) {
+  if (ctx.cert_chain_ == nullptr) {
+    return;
+  }
+
+  auto populate = [&](const std::string& sn) {
+    std::string sn_pattern = sn;
+    if (absl::StartsWith(sn, "*.")) {
+      sn_pattern = sn.substr(1);
+    }
+    PkeyTypesMap pkey_types_map;
+    auto sn_match = server_names_map_.try_emplace(sn_pattern, pkey_types_map).first;
+    auto pt_match = sn_match->second.find(pkey_id);
+    if (pt_match != sn_match->second.end()) {
+      throw EnvoyException(fmt::format(
+          "Failed to load certificate chain from {}, at most one "
+          "certificate of a given type may be specified for each DNS SAN entry or Subject CN: {}",
+          ctx.cert_chain_file_path_, sn_match->first));
+    }
+    sn_match->second.emplace(std::pair<int, std::reference_wrapper<TlsContext>>(pkey_id, ctx));
+  };
+
+  bssl::UniquePtr<GENERAL_NAMES> san_names(static_cast<GENERAL_NAMES*>(
+      X509_get_ext_d2i(ctx.cert_chain_.get(), NID_subject_alt_name, nullptr, nullptr)));
+  auto dns_sans = Utility::getSubjectAltNames(*ctx.cert_chain_, GEN_DNS);
+  // https://www.rfc-editor.org/rfc/rfc6066#section-3
+  // Currently, the only server names supported are DNS hostnames, so we
+  // only save dns san entries to match SNI.
+  for (const auto& san : dns_sans) {
+    populate(san);
+  }
+
+  // https://www.rfc-editor.org/rfc/rfc6125#section-6.4.4
+  // As noted, a client MUST NOT seek a match for a reference identifier
+  // of CN-ID if the presented identifiers include a DNS-ID, SRV-ID,
+  // URI-ID, or any application-specific identifier types supported by the
+  // client.
+  if (san_names.get() == nullptr) {
+    X509_NAME* cert_subject = X509_get_subject_name(ctx.cert_chain_.get());
```

Comment:
I'm questioning whether it's worth having the code for CN at all. Golang clients stopped allowing validation on CN without setting a debug env var to override many releases ago. Can we remove and only use SANs, to simplify code and tests? Note that this wouldn't prevent having CN in a cert, just that Envoy would only use SANs to pick a cert for the client.

Reply:
Golang clients can not cover all use cases. Envoy can be used as gateway or standalone proxy, we can not predict what kind of certs the users will inject into Envoy, and don't know how the client will verify it. I have no idea that how often the CN is used in real world, but to make Envoy can be leveraged in most cases, maybe we can keep it.

Reply:
Golang clients certainly aren't all use cases, but it shows that CN isn't even a viable option if any of the clients for the service are golang. CN has been deprecated for years; the RFCs say to use SANs always, but that CN can be used to support legacy clients.

Removing this functionality wouldn't prevent Envoy from serving certs with a CN, we'd just not support reading the CN for SNI cert selection. In this case, the cert should have a SAN for the value in the CN anyways, so a normal cert would work identically (because there should be a SAN for the CN value).

Reply:
I think we should not ignore legacy clients/servers.

As you know, I proposed TLS bumping, and this feature is one of the dependencies. TLS bumping is very important in  SASE(Secure Access Service Edge), the idea is to decrypt traffic data from client and apply some security policies. For example, we have an employee, we need to make sure he will not make any secret data leak outside.

Thus in TLS bumping, we let clients trust a ca which can be provided by corporate, and we need to dynamically copy the information from original server cert and mimic certs(signed from the ca) for client, in this case, we can not predict if server cert support SAN, if no, the only thing we can leverage is CN. If client support verification based on CN but envoy does not support cert selection based on CN, we get miss match.

I understand that CN is not very often used, but I think it will not break anything, and it's an enhancement to support legacy scenarios.

Reply:
Ok, sounds reasonable.

Please add tests validating that if SANs are present, the CN is ignored.

Reply:
Sure, I have added the test for this case